game-core/src/main/java/games/strategy/engine/framework/startup/ui/GameSelectorPanel.java

```diff
@@ -380,10 +383,12 @@ private void selectGameFile(final boolean saved) {
             }
             model.load(entry);
           });
-          setOriginalPropertiesMap(model.getGameData());
-          // only for new games, not saved games, we set the default options, and set them only once
-          // (the first time it is loaded)
-          gamePropertiesCache.loadCachedGamePropertiesInto(model.getGameData());
+          if (model.getGameData() != null) {
```

Comment:
There's a potential NPE here due to the TOCTOU, similar to what @RoiEXLab fixed in #2997.  I'd suggest storing `model.getGameData()` in a local, and using that in the code that follows.

Reply:
In the error case, we get a `null` gameData which will go on to NPE in the next couple lines without this check.

`model` does not seem like it can be null, i'm not sure I follow where the NPE would be.

Reply:
The `GameData` stored in `model` looks like it can be changed by a different thread.  So the `if` check on line 386 may succeed, but then `model.getGameData()` could return `null` on lines 387 and 390 if another thread sneaks in and calls `model.setGameData(null)`.  That seems to defeat the purpose of the `if` check on line 386.

Reply:
Would the other thread set it to null? 

This is an existing problem to some extent, so I'm wondering if it is appropriate to mark a TODO. I don't think I can/want to solve the sync problem at this moment.

Reply:
It looks like: ```model.setGameData(null)```
is triggered on cancel + shutdown. I think we might be thread safe because both are event driven from EDT.

Looking at GameSelectorModel, there is a synchronize around the set and get methods. In theory those should guard against this. 

@ssoloff do you still see a potential problem? If a threading problem is here, are we in agreement it's an existing issue? It could be there, though I'm not following just yet how removing the extra fallback would do it.

Reply:
The synchronization on `getGameData()` and `setGameData()` only prevents those methods from running concurrently.  It does not prevent a race _between_ calls.  Here's one possible scenario:

```java
if (model.getGameData() != null) { // <-- returns non-null on thread 1
  // <-- context switch to thread 2 which calls model.setGameData(null)
  setOriginalPropertiesMap(model.getGameData()); // <-- context switch back to thread 1; now returns null
  gamePropertiesCache.loadCachedGamePropertiesInto(model.getGameData());
}
```

Yes, there may be an existing issue.  The only reason I'm bringing it up here is because this PR added the `if (model.getGameData() != null)` guard.  I'm assuming adding that guard was important for some reason to ensure the code in the `if` block not be run if `model.getGameData()` returns `null`.  My point is the guard is not sufficient because there is a TOCTOU due to a second thread changing what `model.getGameData()` may return _after_ the `if` check.

There may be no problem _today_.  **But the way the `GameSelectorModel` API is written, callers need to assume subsequent calls to `getGameData()` can return different values.**  A dev in the future will most likely assume `GameSelectorModel` clients respect the API and could introduce a change that will break what "works today."  That's most likely what happened with the issue @ron-murhammer reported in #2980.

All I'm proposing is getting rid of the TOCTOU via a local:

```java
final GameData gameData = model.getGameData();
if (gameData != null) {
  setOriginalPropertiesMap(gameData);
  // only for new games, not saved games, we set the default options, and set them only once
  // (the first time it is loaded)
  gamePropertiesCache.loadCachedGamePropertiesInto(gameData);
}
```

Reply:
> The synchronization on getGameData() and setGameData() only prevents those methods from running concurrently. 

Indeed. I suspect they do not do much good anywhere... We can drop that from this thread : )


> My point is the guard is not sufficient because there is a TOCTOU due to a second thread changing what model.getGameData() may return after the if check.

I think it is the case that when we enter the background thread the UI thread is blocked. The only places where we can insert a null are from actions triggered by UI thread. I'm not 100% sure as this is kinda complex, but I do think we are thread safe in this case as a user cannot initiate any of those other calls that would not be safe, as the UI thread is blocked.

Reply:
@ssoloff are you sure there is a TOCTOU problem? The specific places we null out the game data, unless I'm mistaken they are all driven by UI events. Meanwhile, while loading game data, we have a modal loading dialog. I suspect there is no TOCTOU.

Reply:
I'm saying there's a _potential_ for a TOCTOU given how the `GameSelectorModel` API is defined.  

Regardless, we've spent a bit too much time on an issue that, as you said, is probably not manifest today. :smile: